### PR TITLE
GO-3046: Files: New uploading / removing queue

### PR DESCRIPTION
**Pull request by deff7**
## Summary by CodeRabbit

- **New Features**
    - Introduced new methods and processes for file synchronization, including handling file deletion and upload limits.
    - Added functionality for file object migration, including services and testing for migration processes.
    - Implemented a persistent queue system for managing file synchronization tasks.
    - Updated file status handling to use a new custom status type for improved tracking.
    - Updated counter values and added new reindexing processes for file objects.

- **Bug Fixes**
    - Enhanced error handling across various modules including file storage and identity management.
    - Fixed issues related to file ID validation and synchronization logic.

- **Refactor**
    - Major refactoring of file synchronization logic to streamline operations and improve performance.
    - Adjusted the order of component registration during system bootstrap for optimized initialization.

- **Documentation**
    - Updated test setups and assertions to align with new functionalities and error handling mechanisms.

- **Tests**
    - Expanded test coverage for new functionalities in file handling and synchronization services.
    - Added tests for persistent queue operations and file migration scenarios.

- **Chores**
    - Cleaned up imports and unused variables across several modules to maintain code cleanliness and efficiency.

### CODE REVIEW

**Review by fat-fellow — APPROVED** [review]

File: core/queue/queue.go [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+	"github.com/anyproto/anytype-heart/util/badgerhelper"
+)
+
+var errRemoved = fmt.Errorf("removed from queue")
+
+type Item interface {
+	Key() string
+}
+
+type ItemWithOrder interface {
```
Comment (fat-fellow) [inline core/queue/queue.go:26]: OrderedItem?

File: core/block/editor/smartblock/smartblock.go [hunk 2]
```diff
@@ -1374,7 +1381,8 @@ func (sb *smartBlock) injectDerivedDetails(s *state.State, spaceID string, sbt s
 	}
 
 	if v, ok := s.Details().GetFields()[bundle.RelationKeyFileBackupStatus.String()]; ok {
-		s.SetDetailAndBundledRelation(bundle.RelationKeyFileSyncStatus, v)
+		status := filesyncstatus.Status(v.GetNumberValue())
+		s.SetDetailAndBundledRelation(bundle.RelationKeyFileSyncStatus, pbtypes.Int64(int64(status.ToSyncStatus())))
```
Comment (deff7) [inline core/block/editor/smartblock/smartblock.go:1385]: Comment: Client expects syncstatus constants in this relation

File: core/files/fileobject/service.go [hunk 3]
```diff
@@ -107,15 +109,82 @@ func (s *service) Init(a *app.App) error {
...
+	db, err := dbProvider.LocalStorage()
+	if err != nil {
+		return fmt.Errorf("get badger: %w", err)
+	}
+	s.migrationQueue = queue.New(db, log.Desugar(), []byte("queue/file_migration/"), makeMigrationItem, s.migrationQueueHandler)
 	return nil
 }
 
 func (s *service) Run(_ context.Context) error {
+	go func() {
```
Comment (deff7) [inline core/files/fileobject/service.go:125]: Don't block starting of application

File: core/filestorage/filesync/upload.go [hunk 4]
```diff
@@ -9,25 +9,36 @@ import (
...
+		Imported:    imported,
+		Timestamp:   time.Now().UnixMilli(),
+	}
+	err = it.Validate()
+	if err != nil {
+		return fmt.Errorf("validate queue item: %w", err)
 	}
-	return
+
+	if !f.retryUploadingQueue.Has(it.Key()) && !f.deletionQueue.Has(it.Key()) {
```
Comment (deff7) [inline core/filestorage/filesync/upload.go:38]: Fix: check that file is not in retryDeletionQueue

File: core/queue/queue.go [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+		return nil
+	}
+	q.set[item.Key()] = struct{}{}
+	q.lock.Unlock()
+
+	err = q.batcher.Add(q.ctx, item)
+	if err != nil {
+		return err
+	}
+	return q.store(item)
```
Comment (mcrakhman) [inline core/queue/queue.go:255]: I would split all the badger parts into other components, maybe just provide an interface inside queue which would store the items, and remove them. And make this interface abstract

File: core/queue/queue.go (hunk 1 above)
Comment (deff7) [reply core/queue/queue.go]: fixed

File: core/block/editor/smartblock/smartblock.go (hunk 2 above)
Comment (deff7) [reply core/block/editor/smartblock/smartblock.go:1385]: fixed

**Comment by deff7** [comment]
@coderabbitai review

File: core/filestorage/filesync/queue.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+func (it *QueueItem) FullFileId() domain.FullFileId {
+	return domain.FullFileId{
+		SpaceId: it.SpaceId,
+		FileId:  it.FileId,
+	}
+}
+
+func (it *QueueItem) Less(other *QueueItem) bool {
+	return it.Timestamp < other.Timestamp
+}
```
[the remark this answers is not in the document]
Comment (deff7) [reply core/filestorage/filesync/queue.go:48]: yes, please

File: core/domain/file.go [hunk 7]
```diff
@@ -8,15 +8,23 @@ import (
...
 type FileId string
 
-func (h FileId) String() string {
-	return string(h)
+func (id FileId) String() string {
+	return string(id)
+}
+
+func (id FileId) Valid() bool {
+	return IsFileId(string(id))
```
[the remark this answers is not in the document]
Comment (deff7) [reply core/domain/file.go:16]: yes, try it

File: core/filestorage/filesync/filesync.go [hunk 8]
```diff
@@ -55,29 +52,35 @@
 }
 
 type QueueInfo struct {
-	UploadingQueue []*QueueItem
-	DiscardedQueue []*QueueItem
-	RemovingQueue  []*QueueItem
+	UploadingQueue      []string
+	RetryUploadingQueue []string
+	DeletionQueue       []string
+	RetryDeletionQueue  []string
```
[the remark this answers is not in the document]
Comment (deff7) [reply core/filestorage/filesync/filesync.go:58]: go on

File: core/filestorage/filesync/filesync.go [hunk 9]
```diff
@@ -87,76 +90,94 @@ func New() FileSync {
...
+	s.dbProvider = app.MustComponent[datastore.Datastore](a)
+	s.rpcStore = a.MustComponent(rpcstore.CName).(rpcstore.Service).NewStore()
+	s.dagService = a.MustComponent(fileservice.CName).(fileservice.FileService).DAGService()
+	s.fileStore = app.MustComponent[filestore.FileStore](a)
+	s.eventSender = app.MustComponent[event.Sender](a)
+	s.removePingCh = make(chan struct{})
+	s.uploadPingCh = make(chan struct{})
+	db, err := s.dbProvider.LocalStorage()
+	if err != nil {
+		return
```
Comment (cheggaaa) [inline core/filestorage/filesync/filesync.go:103]: In case of err here fileSync.Close will panic on queues closing. Maybe check for nil in Close

Comment (deff7) [reply core/filestorage/filesync/filesync.go:103]: thank you, fixed

File: core/filestorage/filesync/delete_test.go [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+		err := fx.DeleteFile("objectId", domain.FullFileId{SpaceId: spaceId, FileId: testFileId})
+		require.NoError(t, err)
+
+		fx.waitEmptyQueue(t, fx.deletionQueue, time.Second*1)
+
+		fx.waitCondition(t, 100*time.Millisecond, func() bool {
+			return fx.retryDeletionQueue.Len() == 1 || fx.retryDeletionQueue.NumProcessedItems() > 0
+		})
+	})
+}
```
[the remark this answers is not in the document]
Comment (deff7) [reply core/filestorage/filesync/delete_test.go:47]: try it

**Review by cheggaaa — APPROVED** [review]

**Pull request merged**